File: app/src/main/java/com/developer/fabian/tictactoe/GameActivity.java
package com.developer.fabian.tictactoe;

import android.content.pm.ActivityInfo;
import android.graphics.Color;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import com.developer.fabian.tictactoe.game.TicTacToe;

public class GameActivity extends AppCompatActivity {

    private int humanResult;
    private int androidResult;
    private int tieResult;

    private TicTacToe mGame;

    private Button mBoardButtons[];
    private TextView mInfoTextView;
    private TextView mResultTextView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
        setContentView(R.layout.activity_game);

        mBoardButtons = new Button[TicTacToe.BOARD_SIZE];
        mBoardButtons[0] = findViewById(R.id.one);
        mBoardButtons[1] = findViewById(R.id.two);
        mBoardButtons[2] = findViewById(R.id.three);
        mBoardButtons[3] = findViewById(R.id.four);
        mBoardButtons[4] = findViewById(R.id.five);
        mBoardButtons[5] = findViewById(R.id.six);
        mBoardButtons[6] = findViewById(R.id.seven);
        mBoardButtons[7] = findViewById(R.id.eight);
        mBoardButtons[8] = findViewById(R.id.nine);

        mResultTextView = findViewById(R.id.results);
        mInfoTextView = findViewById(R.id.dialogue);

        mGame = new TicTacToe();

        startNewGame();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        super.onCreateOptionsMenu(menu);
        menu.add(R.string.menu_new_game);

        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        startNewGame();
        return true;
    }

    private void startNewGame() {
        mGame.clearBoard();

        for (int i = 0; i < mBoardButtons.length; i++) {
            mBoardButtons[i].setText("");
            mBoardButtons[i].setEnabled(true);
            mBoardButtons[i].setOnClickListener(new ButtonClickListener(i));
        }

        mInfoTextView.setText(R.string.first_human);
        mResultTextView.setText(String.format(getString(R.string.result_text), humanResult, tieResult, androidResult));
    }

    private void setMove(char player, int location) {
        mGame.setMove(player, location);
        mBoardButtons[location].setEnabled(false);
        mBoardButtons[location].setText(String.valueOf(player));

        if (player == TicTacToe.HUMAN_PLAYER)
            mBoardButtons[location].setTextColor(Color.rgb(0, 200, 0));
        else
            mBoardButtons[location].setTextColor(Color.rgb(200, 0, 0));
    }

    private void disableButtons() {
        for (Button mBoardButton : mBoardButtons)
            mBoardButton.setEnabled(false);
    }

    private class ButtonClickListener implements View.OnClickListener {
        int location;

        ButtonClickListener(int location) {
            this.location = location;
        }

        public void onClick(View view) {
            if (mBoardButtons[location].isEnabled()) {
                setMove(TicTacToe.HUMAN_PLAYER, location);

                int winner = mGame.checkForWinner();
                if (winner == 0) {
                    mInfoTextView.setText(R.string.turn_computer);

                    int move = mGame.getComputerMove();
                    setMove(TicTacToe.COMPUTER_PLAYER, move);

                    winner = mGame.checkForWinner();
                }

                if (winner == 0) {
                    mInfoTextView.setText(R.string.turn_human);
                } else if (winner == 1) {
                    mInfoTextView.setText(R.string.result_tie);
                    tieResult++;
                    disableButtons();
                } else if (winner == 2) {
                    mInfoTextView.setText(R.string.result_human_wins);
                    humanResult++;
                    disableButtons();
                } else {
                    mInfoTextView.setText(R.string.result_computer_wins);
                    androidResult++;
                    disableButtons();
                }

                mResultTextView.setText(String.format(getString(R.string.result_text), humanResult, tieResult, androidResult));
            }
        }
    }
}

File: app/src/main/java/com/developer/fabian/tictactoe/GameActivity.kt
package com.developer.fabian.tictactoe

import android.content.pm.ActivityInfo
import android.graphics.Color
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Button
import android.widget.TextView

import com.developer.fabian.tictactoe.game.TicTacToe

class GameActivity : AppCompatActivity() {

    private var humanResult: Int = 0
    private var androidResult: Int = 0
    private var tieResult: Int = 0

    private var mGame: TicTacToe? = null

    private lateinit var mBoardButtons: Array<Button>
    private lateinit var mInfoTextView: TextView
    private lateinit var mResultTextView: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
        setContentView(R.layout.activity_game)

        mBoardButtons = arrayOf(
                findViewById(R.id.one),
                findViewById(R.id.two),
                findViewById(R.id.three),
                findViewById(R.id.four),
                findViewById(R.id.five),
                findViewById(R.id.six),
                findViewById(R.id.seven),
                findViewById(R.id.eight),
                findViewById(R.id.nine)
        )

        mResultTextView = findViewById(R.id.results)
        mInfoTextView = findViewById(R.id.dialogue)

        mGame = TicTacToe()

        startNewGame()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        super.onCreateOptionsMenu(menu)
        menu.add(R.string.menu_new_game)

        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        startNewGame()
        return true
    }

    private fun startNewGame() {
        mGame!!.clearBoard()

        for (i in mBoardButtons.indices) {
            mBoardButtons[i].text = ""
            mBoardButtons[i].isEnabled = true
            mBoardButtons[i].setOnClickListener(ButtonClickListener(i))
        }

        mInfoTextView.setText(R.string.first_human)
        mResultTextView.text = String.format(getString(R.string.result_text), humanResult, tieResult, androidResult)
    }

    private fun setMove(player: Char, location: Int) {
        mGame!!.setMove(player, location)
        mBoardButtons[location].isEnabled = false
        mBoardButtons[location].text = player.toString()

        if (player == TicTacToe.HUMAN_PLAYER)
            mBoardButtons[location].setTextColor(Color.rgb(0, 200, 0))
        else
            mBoardButtons[location].setTextColor(Color.rgb(200, 0, 0))
    }

    private fun disableButtons() {
        for (mBoardButton in mBoardButtons)
            mBoardButton.isEnabled = false
    }

    private inner class ButtonClickListener internal constructor(internal var location: Int) : View.OnClickListener {

        override fun onClick(view: View) {
            if (mBoardButtons[location].isEnabled) {
                setMove(TicTacToe.HUMAN_PLAYER, location)

                var winner = mGame!!.checkForWinner()
                if (winner == 0) {
                    mInfoTextView.setText(R.string.turn_computer)

                    val move = mGame!!.computerMove
                    setMove(TicTacToe.COMPUTER_PLAYER, move)

                    winner = mGame!!.checkForWinner()
                }

                when (winner) {
                    0 -> mInfoTextView.setText(R.string.turn_human)
                    1 -> {
                        mInfoTextView.setText(R.string.result_tie)
                        tieResult++
                        disableButtons()
                    }
                    2 -> {
                        mInfoTextView.setText(R.string.result_human_wins)
                        humanResult++
                        disableButtons()
                    }
                    else -> {
                        mInfoTextView.setText(R.string.result_computer_wins)
                        androidResult++
                        disableButtons()
                    }
                }

                mResultTextView.text = String.format(getString(R.string.result_text), humanResult, tieResult, androidResult)
            }
        }
    }
}
